### PR TITLE
Move sygus grammar utilities to separate file.

### PR DESCRIPTION
**Pull request by ajreynol**
This moves all utilities for generating inductive datatypes corresponding to SyGuS grammars into a separate file.

It also changes a few functions from recursive to iterative that were carried over to the new class CegGrammarCons

### CODE REVIEW

File: src/theory/quantifiers/term_database.h [hunk 1]
```diff
@@ -19,6 +19,8 @@
 
 #include <map>
 #include <unordered_set>
+#include <unordered_map>
+#include <stack>
```
Comment (mpreiner) [inline src/theory/quantifiers/term_database.h]: These #includes should be placed in the .cpp were they are needed.

File: src/theory/quantifiers_engine.h [hunk 2]
```diff
@@ -20,6 +20,8 @@
 #include <iostream>
 #include <map>
 #include <unordered_map>
+#include <unordered_set>
+#include <stack>
```
Comment (mpreiner) [inline src/theory/quantifiers_engine.h]: Ditto.

**Comment by ajreynol** [comment]
I moved the includes for now.  

There are still times when I'd like to use std::map instead of std::unordered_map.  The new code "templates" and "templates_arg" is an example, where the order in the map keeps the order in the quantifier prefix (although this is not critical, this is nice to have).

**Comment by mpreiner** [comment]
Well in this case you wanted them to keep the order and  the use of std::map is OK.

**Pull request merged**